ext/standard/math.c

```diff
@@ -93,27 +93,56 @@ static inline double php_intpow10(int power) {
 /* {{{ php_round_helper
        Actually performs the rounding of a value to integer in a certain mode */
 static inline double php_round_helper(double value, int mode) {
-	double tmp_value;
+	double integral, fractional;
 
-	if (value >= 0.0) {
-		tmp_value = floor(value + 0.5);
-		if ((mode == PHP_ROUND_HALF_DOWN && value == (-0.5 + tmp_value)) ||
-			(mode == PHP_ROUND_HALF_EVEN && value == (0.5 + 2 * floor(tmp_value/2.0))) ||
-			(mode == PHP_ROUND_HALF_ODD  && value == (0.5 + 2 * floor(tmp_value/2.0) - 1.0)))
-		{
-			tmp_value = tmp_value - 1.0;
-		}
-	} else {
-		tmp_value = ceil(value - 0.5);
-		if ((mode == PHP_ROUND_HALF_DOWN && value == (0.5 + tmp_value)) ||
-			(mode == PHP_ROUND_HALF_EVEN && value == (-0.5 + 2 * ceil(tmp_value/2.0))) ||
-			(mode == PHP_ROUND_HALF_ODD  && value == (-0.5 + 2 * ceil(tmp_value/2.0) + 1.0)))
-		{
-			tmp_value = tmp_value + 1.0;
-		}
+	fractional = fabs(modf(value, &integral));
+
+	switch (mode) {
```

Comment:
the code can be simplified, as for <0.5 and >0.5 the code is the same, also every return is the same

so the switch can be replaced by:
```
if (fractional > 0.5) {
    return integral + copysign(1.0, integral);
} else if (fractional == 0.5) {
    if (mode == PHP_ROUND_HALF_UP) {
        return integral + copysign(1.0, integral);
    } else if (mode == PHP_ROUND_HALF_EVEN || mode == PHP_ROUND_HALF_ODD) {
        bool even = !fmod(integral, 2.0);
        if (even == (mode == PHP_ROUND_HALF_ODD)) {
            return integral + copysign(1.0, integral);
        }
    }
}
```

Reply:
The suggested code is horrendous to read compared to the ``switch`` statement version.

On top of that I would imagine a compiler to produce more efficient code via jump tables with a switch statement then those nested if/else blocks.

Reply:
The code should be the fastest for `fractional != 0.5`.

Reply:
It's not.